aiida_common_workflows/workflows/bands/castep/generator.py

```diff
@@ -0,0 +1,71 @@
+
+from aiida_common_workflows.generators import CodeType
+
+from ..generator import CommonBandsInputGenerator
+
+__all__ = ('CastepCommonBandsInputGenerator',)
+
+
+class CastepCommonBandsInputGenerator(CommonBandsInputGenerator):
+    """Generator of inputs for the CastepCommonBandsWorkChain"""
+
+    @classmethod
+    def define(cls, spec):
+        """Define the specification of the input generator.
+
+        The ports defined on the specification are the inputs that will be accepted by the ``get_builder`` method.
+        """
+        super().define(spec)
+        spec.inputs['engines']['bands']['code'].valid_type = CodeType('castep.castep')
+
+    def _construct_builder(self, **kwargs) -> engine.ProcessBuilder:
+        """Construct a process builder based on the provided keyword arguments.
+
+        The keyword arguments will have been validated against the input generator specification.
+        """
+        # pylint: disable=too-many-branches,too-many-statements,too-many-locals
+        engines = kwargs.get('engines', None)
+        parent_folder = kwargs['parent_folder']
+        bands_kpoints = kwargs['bands_kpoints']
+
+        # From the parent folder, we retrieve the calculation that created it. Note
+        # that we are sure it exists (it wouldn't be the same for WorkChains). We then check
+        # that it is a CastepCalculation and create the builder.
+        parent_castep_calc = parent_folder.creator
+        if parent_castep_calc.process_type != 'aiida.calculations:castep.castep':
+            raise ValueError('The `parent_folder` has not been created by a CastepCalculation')
+        builder_castep_calc = parent_castep_calc.get_builder_restart()
+
+        # Construct the builder of the `common_bands_wc` from the builder of a CastepCalculation.
+        builder_common_bands_wc = self.process_class.get_builder()
+        builder_common_bands_wc.scf.calc_options = orm.Dict(dict=dict(builder_castep_calc.metadata.options))
+        # Ensure we use castep_bin for restart, instead of the check file
+        #builder_common_bands_wc.scf.options = orm.Dict(dict={'use_castep_bin': True})
+
+        builder_castep_calc.metadata = {}
+
+        # Attach inputs of the calculation
+        for key, value in builder_castep_calc.items():
+            if value and key not in ['metadata', 'structure']:
+                builder_common_bands_wc.scf.calc[key] = value
+
+        # Updated the structure (in case we have one in output)
+        if 'output_structure' in parent_castep_calc.outputs:
+            builder_common_bands_wc.structure = parent_castep_calc.outputs.output_structure
+        else:
+            builder_common_bands_wc.structure = parent_castep_calc.inputs.structure
+
+        engb = engines['bands']
+        builder_common_bands_wc.scf.calc.code = engines['bands']['code']
+        if 'options' in engb:
+            builder_common_bands_wc.scf.calc_options = orm.Dict(dict=engines['bands']['options'])
```

Comment:
I think you may have just copied this from the Siesta implementation. Does the Castep implementation also expect the options as a `Dict` node as opposed to just a dictionary in the `metadata` namespace? If so, I would say that is suboptimal and we should think of improving that in `aiida-castep` at some point.

Reply:
CASTEP actually work with `metadata` exposed from `CalcJob` previously, e.g. one pass the `metadata.options` as a `dict` not `Dict`, to an exposed `calc.metadata` port namespace. 

I just updated the code to make it work with `Dict`...

Which way do you think is better? For the plain `metadata.options` approach the problem I had is that one cannot just do `get_builder_restart` from a `WorkChainNode` and have the builder fully populated and ready to go. This is because the exposed `metadata.options` namespace remains empty.